{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "37a6dd7c",
   "metadata": {
    "_cell_guid": "b1076dfc-b9ad-4769-8c92-a6c4dae69d19",
    "_uuid": "8f2839f25d086af736a60e9eeb907d3b93b6e0e5",
    "execution": {
     "iopub.execute_input": "2022-08-10T12:04:58.672892Z",
     "iopub.status.busy": "2022-08-10T12:04:58.671729Z",
     "iopub.status.idle": "2022-08-10T12:05:11.996201Z",
     "shell.execute_reply": "2022-08-10T12:05:11.995054Z",
     "shell.execute_reply.started": "2022-08-10T11:26:29.048911Z"
    },
    "papermill": {
     "duration": 13.333988,
     "end_time": "2022-08-10T12:05:11.996406",
     "exception": false,
     "start_time": "2022-08-10T12:04:58.662418",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\r\n",
      "tensorflow-cloud 0.1.13 requires tensorflow-datasets<3.1.0, but you have tensorflow-datasets 4.4.0 which is incompatible.\u001b[0m\r\n",
      "\u001b[33mWARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "!pip install -q tensorflow_datasets==4.4.0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "188b5267",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-08-10T12:05:12.023007Z",
     "iopub.status.busy": "2022-08-10T12:05:12.022261Z",
     "iopub.status.idle": "2022-08-10T12:05:13.898627Z",
     "shell.execute_reply": "2022-08-10T12:05:13.897849Z",
     "shell.execute_reply.started": "2022-08-10T11:40:46.660690Z"
    },
    "papermill": {
     "duration": 1.893686,
     "end_time": "2022-08-10T12:05:13.898789",
     "exception": false,
     "start_time": "2022-08-10T12:05:12.005103",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cloning into '/kaggle/working/repo'...\r\n",
      "remote: Enumerating objects: 166, done.\u001b[K\r\n",
      "remote: Counting objects: 100% (13/13), done.\u001b[K\r\n",
      "remote: Compressing objects: 100% (12/12), done.\u001b[K\r\n",
      "remote: Total 166 (delta 3), reused 0 (delta 0), pack-reused 153\u001b[K\r\n",
      "Receiving objects: 100% (166/166), 43.44 KiB | 988.00 KiB/s, done.\r\n",
      "Resolving deltas: 100% (53/53), done.\r\n"
     ]
    }
   ],
   "source": [
    "!git clone https://github.com/dleqhuy/cbis_ddsm_train.git /kaggle/working/repo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "985f75ee",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-08-10T12:05:13.918617Z",
     "iopub.status.busy": "2022-08-10T12:05:13.917801Z",
     "iopub.status.idle": "2022-08-10T12:05:13.920606Z",
     "shell.execute_reply": "2022-08-10T12:05:13.921260Z",
     "shell.execute_reply.started": "2022-08-10T11:40:48.181417Z"
    },
    "papermill": {
     "duration": 0.015928,
     "end_time": "2022-08-10T12:05:13.921447",
     "exception": false,
     "start_time": "2022-08-10T12:05:13.905519",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "os.chdir('/kaggle/working/repo/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da31dc8d",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-08-10T12:05:13.939655Z",
     "iopub.status.busy": "2022-08-10T12:05:13.938634Z",
     "iopub.status.idle": "2022-08-10T12:19:25.123942Z",
     "shell.execute_reply": "2022-08-10T12:19:25.123084Z"
    },
    "papermill": {
     "duration": 851.195872,
     "end_time": "2022-08-10T12:19:25.124230",
     "exception": false,
     "start_time": "2022-08-10T12:05:13.928358",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-08-10 12:05:15.818305: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/conda/lib\r\n",
      "2022-08-10 12:05:15.818409: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\r\n",
      "2022-08-10 12:05:23.325951: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set\r\n",
      "2022-08-10 12:05:23.328811: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/conda/lib\r\n",
      "2022-08-10 12:05:23.328860: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)\r\n",
      "2022-08-10 12:05:23.328890: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (7b4cc419d731): /proc/driver/nvidia/version does not exist\r\n",
      "2022-08-10 12:05:23.331407: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\r\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\r\n",
      "2022-08-10 12:05:23.333033: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set\r\n",
      "2022-08-10 12:05:23.336786: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set\r\n",
      "2022-08-10 12:05:23.366442: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job worker -> {0 -> 10.0.0.2:8470}\r\n",
      "2022-08-10 12:05:23.366667: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job localhost -> {0 -> localhost:30064}\r\n",
      "2022-08-10 12:05:23.386176: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job worker -> {0 -> 10.0.0.2:8470}\r\n",
      "2022-08-10 12:05:23.386253: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job localhost -> {0 -> localhost:30064}\r\n",
      "2022-08-10 12:05:23.388225: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:411] Started server with target: grpc://localhost:30064\r\n",
      "I0810 12:05:23.389172 139973921728320 remote.py:218] Entering into master device scope: /job:worker/replica:0/task:0/device:CPU:0\r\n",
      "I0810 12:05:23.390082 139973921728320 tpu_strategy_util.py:74] Initializing the TPU system: grpc://10.0.0.2:8470\r\n",
      "I0810 12:05:29.274822 139973921728320 tpu_strategy_util.py:109] Clearing out eager caches\r\n",
      "I0810 12:05:29.283546 139973921728320 tpu_strategy_util.py:133] Finished initializing TPU system.\r\n",
      "I0810 12:05:29.287990 139973921728320 tpu_system_metadata.py:159] Found TPU system:\r\n",
      "I0810 12:05:29.288267 139973921728320 tpu_system_metadata.py:160] *** Num TPU Cores: 8\r\n",
      "I0810 12:05:29.288796 139973921728320 tpu_system_metadata.py:161] *** Num TPU Workers: 1\r\n",
      "I0810 12:05:29.288932 139973921728320 tpu_system_metadata.py:163] *** Num TPU Cores Per Worker: 8\r\n",
      "I0810 12:05:29.289058 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:localhost/replica:0/task:0/device:CPU:0, CPU, 0, 0)\r\n",
      "I0810 12:05:29.289347 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:CPU:0, CPU, 0, 0)\r\n",
      "I0810 12:05:29.289476 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:0, TPU, 0, 0)\r\n",
      "I0810 12:05:29.289621 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:1, TPU, 0, 0)\r\n",
      "I0810 12:05:29.289734 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:2, TPU, 0, 0)\r\n",
      "I0810 12:05:29.289854 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:3, TPU, 0, 0)\r\n",
      "I0810 12:05:29.289951 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:4, TPU, 0, 0)\r\n",
      "I0810 12:05:29.290069 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:5, TPU, 0, 0)\r\n",
      "I0810 12:05:29.290163 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:6, TPU, 0, 0)\r\n",
      "I0810 12:05:29.290278 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:7, TPU, 0, 0)\r\n",
      "I0810 12:05:29.290394 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU_SYSTEM:0, TPU_SYSTEM, 0, 0)\r\n",
      "I0810 12:05:29.290511 139973921728320 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:XLA_CPU:0, XLA_CPU, 0, 0)\r\n",
      "2022-08-10 12:05:29.682259: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:29.758817: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:29.830644: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:29.898873: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:29.968344: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.041328: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.108065: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.181291: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.247926: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.381818: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.459390: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "I0810 12:05:30.535290 139973921728320 dataset_info.py:358] Load dataset info from gs://kds-0289b529d96a785933aaa9a1abf3af61a5788358dc2bbd021561b2ef/full_images/1.0.0\r\n",
      "2022-08-10 12:05:30.584377: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.719250: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.794544: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.906570: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:30.983677: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:31.052472: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:05:31.123278: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "I0810 12:05:31.199337 139973921728320 dataset_builder.py:351] Reusing dataset full_images (gs://kds-0289b529d96a785933aaa9a1abf3af61a5788358dc2bbd021561b2ef/full_images/1.0.0)\r\n",
      "I0810 12:05:31.199915 139973921728320 logging_logger.py:36] Constructing tf.data.Dataset full_images for split ['train', 'test', 'val'], from gs://kds-0289b529d96a785933aaa9a1abf3af61a5788358dc2bbd021561b2ef/full_images/1.0.0\r\n",
      "2022-08-10 12:05:31.200228: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "I0810 12:05:31.843448 139973921728320 train_full.py:153] Huấn luyện giai đoạn 1 với các lớp được thêm vào\r\n",
      "Model: \"model_1\"\r\n",
      "_________________________________________________________________\r\n",
      "Layer (type)                 Output Shape              Param #   \r\n",
      "=================================================================\r\n",
      "input_1 (InputLayer)         [(None, 448, 448, 3)]     0         \r\n",
      "_________________________________________________________________\r\n",
      "model (Functional)           (None, None, None, 512)   14714688  \r\n",
      "_________________________________________________________________\r\n",
      "sequential_3 (Sequential)    (None, 7, 7, 256)         1179904   \r\n",
      "_________________________________________________________________\r\n",
      "sequential_4 (Sequential)    (None, 3, 3, 256)         590080    \r\n",
      "_________________________________________________________________\r\n",
      "global_average_pooling2d (Gl (None, 256)               0         \r\n",
      "_________________________________________________________________\r\n",
      "dense (Dense)                (None, 1)                 257       \r\n",
      "=================================================================\r\n",
      "Total params: 16,484,929\r\n",
      "Trainable params: 16,484,929\r\n",
      "Non-trainable params: 0\r\n",
      "_________________________________________________________________\r\n",
      "I0810 12:05:34.870419 139973921728320 train_full.py:165] None\r\n",
      "Epoch 1/50\r\n",
      "184/184 [==============================] - 36s 197ms/step - loss: 0.8795 - accuracy: 0.5224 - val_loss: 0.7095 - val_accuracy: 0.5147\r\n",
      "\r\n",
      "Epoch 00001: val_accuracy improved from -inf to 0.51471, saving model to ./vgg16.h5\r\n",
      "Epoch 2/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.7229 - accuracy: 0.5217 - val_loss: 0.7279 - val_accuracy: 0.5037\r\n",
      "\r\n",
      "Epoch 00002: val_accuracy did not improve from 0.51471\r\n",
      "Epoch 3/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.7004 - accuracy: 0.5340 - val_loss: 0.7034 - val_accuracy: 0.5404\r\n",
      "\r\n",
      "Epoch 00003: val_accuracy improved from 0.51471 to 0.54044, saving model to ./vgg16.h5\r\n",
      "Epoch 4/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6935 - accuracy: 0.5306 - val_loss: 0.7308 - val_accuracy: 0.5074\r\n",
      "\r\n",
      "Epoch 00004: val_accuracy did not improve from 0.54044\r\n",
      "Epoch 5/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6916 - accuracy: 0.5346 - val_loss: 0.7495 - val_accuracy: 0.5441\r\n",
      "\r\n",
      "Epoch 00005: val_accuracy improved from 0.54044 to 0.54412, saving model to ./vgg16.h5\r\n",
      "Epoch 6/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6856 - accuracy: 0.5639 - val_loss: 0.7487 - val_accuracy: 0.5184\r\n",
      "\r\n",
      "Epoch 00006: val_accuracy did not improve from 0.54412\r\n",
      "Epoch 7/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6805 - accuracy: 0.5584 - val_loss: 0.7067 - val_accuracy: 0.5404\r\n",
      "\r\n",
      "Epoch 00007: val_accuracy did not improve from 0.54412\r\n",
      "Epoch 8/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6807 - accuracy: 0.5632 - val_loss: 0.7235 - val_accuracy: 0.5294\r\n",
      "\r\n",
      "Epoch 00008: val_accuracy did not improve from 0.54412\r\n",
      "Epoch 9/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6796 - accuracy: 0.5591 - val_loss: 0.7324 - val_accuracy: 0.5551\r\n",
      "\r\n",
      "Epoch 00009: val_accuracy improved from 0.54412 to 0.55515, saving model to ./vgg16.h5\r\n",
      "Epoch 10/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6828 - accuracy: 0.5645 - val_loss: 0.7408 - val_accuracy: 0.5257\r\n",
      "\r\n",
      "Epoch 00010: val_accuracy did not improve from 0.55515\r\n",
      "Epoch 11/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6678 - accuracy: 0.5720 - val_loss: 0.7125 - val_accuracy: 0.5588\r\n",
      "\r\n",
      "Epoch 00011: val_accuracy improved from 0.55515 to 0.55882, saving model to ./vgg16.h5\r\n",
      "Epoch 12/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6839 - accuracy: 0.5713 - val_loss: 0.6958 - val_accuracy: 0.5588\r\n",
      "\r\n",
      "Epoch 00012: val_accuracy did not improve from 0.55882\r\n",
      "Epoch 13/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6721 - accuracy: 0.5795 - val_loss: 0.7326 - val_accuracy: 0.5846\r\n",
      "\r\n",
      "Epoch 00013: val_accuracy improved from 0.55882 to 0.58456, saving model to ./vgg16.h5\r\n",
      "Epoch 14/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6663 - accuracy: 0.6012 - val_loss: 0.7880 - val_accuracy: 0.5221\r\n",
      "\r\n",
      "Epoch 00014: val_accuracy did not improve from 0.58456\r\n",
      "Epoch 15/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6643 - accuracy: 0.5734 - val_loss: 0.7375 - val_accuracy: 0.5662\r\n",
      "\r\n",
      "Epoch 00015: val_accuracy did not improve from 0.58456\r\n",
      "Epoch 16/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6705 - accuracy: 0.5815 - val_loss: 0.6914 - val_accuracy: 0.5882\r\n",
      "\r\n",
      "Epoch 00016: val_accuracy improved from 0.58456 to 0.58824, saving model to ./vgg16.h5\r\n",
      "Epoch 17/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6680 - accuracy: 0.5978 - val_loss: 0.7461 - val_accuracy: 0.5331\r\n",
      "\r\n",
      "Epoch 00017: val_accuracy did not improve from 0.58824\r\n",
      "Epoch 18/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6521 - accuracy: 0.6094 - val_loss: 0.7252 - val_accuracy: 0.5882\r\n",
      "\r\n",
      "Epoch 00018: val_accuracy did not improve from 0.58824\r\n",
      "Epoch 19/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6570 - accuracy: 0.6162 - val_loss: 0.7212 - val_accuracy: 0.6029\r\n",
      "\r\n",
      "Epoch 00019: val_accuracy improved from 0.58824 to 0.60294, saving model to ./vgg16.h5\r\n",
      "Epoch 20/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6607 - accuracy: 0.5951 - val_loss: 0.7425 - val_accuracy: 0.5735\r\n",
      "\r\n",
      "Epoch 00020: val_accuracy did not improve from 0.60294\r\n",
      "Epoch 21/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6513 - accuracy: 0.6101 - val_loss: 0.6993 - val_accuracy: 0.6140\r\n",
      "\r\n",
      "Epoch 00021: val_accuracy improved from 0.60294 to 0.61397, saving model to ./vgg16.h5\r\n",
      "Epoch 22/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6532 - accuracy: 0.6026 - val_loss: 0.7410 - val_accuracy: 0.5809\r\n",
      "\r\n",
      "Epoch 00022: val_accuracy did not improve from 0.61397\r\n",
      "Epoch 23/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6607 - accuracy: 0.5910 - val_loss: 0.7019 - val_accuracy: 0.5735\r\n",
      "\r\n",
      "Epoch 00023: val_accuracy did not improve from 0.61397\r\n",
      "Epoch 24/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6422 - accuracy: 0.6223 - val_loss: 0.6840 - val_accuracy: 0.6103\r\n",
      "\r\n",
      "Epoch 00024: val_accuracy did not improve from 0.61397\r\n",
      "Epoch 25/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6476 - accuracy: 0.6087 - val_loss: 0.7800 - val_accuracy: 0.5368\r\n",
      "\r\n",
      "Epoch 00025: val_accuracy did not improve from 0.61397\r\n",
      "Epoch 26/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6552 - accuracy: 0.5971 - val_loss: 0.7038 - val_accuracy: 0.5662\r\n",
      "\r\n",
      "Epoch 00026: val_accuracy did not improve from 0.61397\r\n",
      "Epoch 27/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6491 - accuracy: 0.6087 - val_loss: 0.6967 - val_accuracy: 0.6029\r\n",
      "\r\n",
      "Epoch 00027: val_accuracy did not improve from 0.61397\r\n",
      "Epoch 28/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6394 - accuracy: 0.6230 - val_loss: 0.6706 - val_accuracy: 0.6287\r\n",
      "\r\n",
      "Epoch 00028: val_accuracy improved from 0.61397 to 0.62868, saving model to ./vgg16.h5\r\n",
      "Epoch 29/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6434 - accuracy: 0.6311 - val_loss: 0.7127 - val_accuracy: 0.5956\r\n",
      "\r\n",
      "Epoch 00029: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 30/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6431 - accuracy: 0.6236 - val_loss: 0.7073 - val_accuracy: 0.6029\r\n",
      "\r\n",
      "Epoch 00030: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 31/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6368 - accuracy: 0.6420 - val_loss: 0.6702 - val_accuracy: 0.6654\r\n",
      "\r\n",
      "Epoch 00031: val_accuracy improved from 0.62868 to 0.66544, saving model to ./vgg16.h5\r\n",
      "Epoch 32/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6285 - accuracy: 0.6420 - val_loss: 0.6861 - val_accuracy: 0.6324\r\n",
      "\r\n",
      "Epoch 00032: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 33/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6393 - accuracy: 0.6311 - val_loss: 0.7374 - val_accuracy: 0.5772\r\n",
      "\r\n",
      "Epoch 00033: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 34/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6219 - accuracy: 0.6406 - val_loss: 0.6577 - val_accuracy: 0.6324\r\n",
      "\r\n",
      "Epoch 00034: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 35/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6280 - accuracy: 0.6311 - val_loss: 0.6529 - val_accuracy: 0.6618\r\n",
      "\r\n",
      "Epoch 00035: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 36/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6153 - accuracy: 0.6454 - val_loss: 0.6417 - val_accuracy: 0.6140\r\n",
      "\r\n",
      "Epoch 00036: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 37/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6347 - accuracy: 0.6529 - val_loss: 0.6580 - val_accuracy: 0.6176\r\n",
      "\r\n",
      "Epoch 00037: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 38/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6155 - accuracy: 0.6522 - val_loss: 0.6465 - val_accuracy: 0.6581\r\n",
      "\r\n",
      "Epoch 00038: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 39/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6175 - accuracy: 0.6569 - val_loss: 0.6485 - val_accuracy: 0.6324\r\n",
      "\r\n",
      "Epoch 00039: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 40/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6258 - accuracy: 0.6508 - val_loss: 0.6549 - val_accuracy: 0.6544\r\n",
      "\r\n",
      "Epoch 00040: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 41/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6088 - accuracy: 0.6569 - val_loss: 0.6603 - val_accuracy: 0.6434\r\n",
      "\r\n",
      "Epoch 00041: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 42/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6151 - accuracy: 0.6501 - val_loss: 0.6079 - val_accuracy: 0.6397\r\n",
      "\r\n",
      "Epoch 00042: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 43/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6100 - accuracy: 0.6522 - val_loss: 0.6467 - val_accuracy: 0.6287\r\n",
      "\r\n",
      "Epoch 00043: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 44/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6148 - accuracy: 0.6488 - val_loss: 0.6503 - val_accuracy: 0.6176\r\n",
      "\r\n",
      "Epoch 00044: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 45/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6040 - accuracy: 0.6610 - val_loss: 0.6613 - val_accuracy: 0.6140\r\n",
      "\r\n",
      "Epoch 00045: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 46/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6024 - accuracy: 0.6766 - val_loss: 0.5921 - val_accuracy: 0.7059\r\n",
      "\r\n",
      "Epoch 00046: val_accuracy improved from 0.66544 to 0.70588, saving model to ./vgg16.h5\r\n",
      "Epoch 47/50\r\n",
      "184/184 [==============================] - 5s 26ms/step - loss: 0.6122 - accuracy: 0.6658 - val_loss: 0.5994 - val_accuracy: 0.6875\r\n",
      "\r\n",
      "Epoch 00047: val_accuracy did not improve from 0.70588\r\n",
      "Epoch 48/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6006 - accuracy: 0.6726 - val_loss: 0.6117 - val_accuracy: 0.6324\r\n",
      "\r\n",
      "Epoch 00048: val_accuracy did not improve from 0.70588\r\n",
      "Epoch 49/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6065 - accuracy: 0.6671 - val_loss: 0.6064 - val_accuracy: 0.6691\r\n",
      "\r\n",
      "Epoch 00049: val_accuracy did not improve from 0.70588\r\n",
      "Epoch 50/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5914 - accuracy: 0.6773 - val_loss: 0.6598 - val_accuracy: 0.6250\r\n",
      "\r\n",
      "Epoch 00050: val_accuracy did not improve from 0.70588\r\n",
      "I0810 12:10:39.571568 139973921728320 train_full.py:176] Mở đóng băng mô hình, huấn luyện giai đoạn 2\r\n",
      "Epoch 50/150\r\n",
      "184/184 [==============================] - 44s 237ms/step - loss: 0.6241 - accuracy: 0.6495 - val_loss: 0.5714 - val_accuracy: 0.6875\r\n",
      "\r\n",
      "Epoch 00050: val_accuracy did not improve from 0.70588\r\n",
      "Epoch 51/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.5978 - accuracy: 0.6753 - val_loss: 0.5893 - val_accuracy: 0.7132\r\n",
      "\r\n",
      "Epoch 00051: val_accuracy improved from 0.70588 to 0.71324, saving model to ./vgg16.h5\r\n",
      "Epoch 52/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.5974 - accuracy: 0.6739 - val_loss: 0.5893 - val_accuracy: 0.6728\r\n",
      "\r\n",
      "Epoch 00052: val_accuracy did not improve from 0.71324\r\n",
      "Epoch 53/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.5839 - accuracy: 0.7018 - val_loss: 0.5222 - val_accuracy: 0.7610\r\n",
      "\r\n",
      "Epoch 00053: val_accuracy improved from 0.71324 to 0.76103, saving model to ./vgg16.h5\r\n",
      "Epoch 54/150\r\n",
      "184/184 [==============================] - 10s 54ms/step - loss: 0.5652 - accuracy: 0.7018 - val_loss: 0.5592 - val_accuracy: 0.7169\r\n",
      "\r\n",
      "Epoch 00054: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 55/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.5847 - accuracy: 0.6821 - val_loss: 0.5428 - val_accuracy: 0.7426\r\n",
      "\r\n",
      "Epoch 00055: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 56/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.5624 - accuracy: 0.7113 - val_loss: 0.5517 - val_accuracy: 0.7390\r\n",
      "\r\n",
      "Epoch 00056: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 57/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.5687 - accuracy: 0.6943 - val_loss: 0.5379 - val_accuracy: 0.7169\r\n",
      "\r\n",
      "Epoch 00057: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 58/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.5462 - accuracy: 0.7208 - val_loss: 0.5310 - val_accuracy: 0.7426\r\n",
      "\r\n",
      "Epoch 00058: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 59/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.5555 - accuracy: 0.7099 - val_loss: 0.5285 - val_accuracy: 0.7463\r\n",
      "\r\n",
      "Epoch 00059: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 60/150\r\n",
      "184/184 [==============================] - 10s 53ms/step - loss: 0.5427 - accuracy: 0.7249 - val_loss: 0.5544 - val_accuracy: 0.7096\r\n",
      "\r\n",
      "Epoch 00060: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 61/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.5438 - accuracy: 0.7385 - val_loss: 0.5535 - val_accuracy: 0.7279\r\n",
      "\r\n",
      "Epoch 00061: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 62/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.5333 - accuracy: 0.7262 - val_loss: 0.5486 - val_accuracy: 0.7316\r\n",
      "\r\n",
      "Epoch 00062: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 63/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.5277 - accuracy: 0.7215 - val_loss: 0.5631 - val_accuracy: 0.7279\r\n",
      "\r\n",
      "Epoch 00063: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 64/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.5186 - accuracy: 0.7391 - val_loss: 0.5333 - val_accuracy: 0.7243\r\n",
      "\r\n",
      "Epoch 00064: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 65/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.5011 - accuracy: 0.7602 - val_loss: 0.6098 - val_accuracy: 0.6765\r\n",
      "\r\n",
      "Epoch 00065: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 66/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.5118 - accuracy: 0.7480 - val_loss: 0.5369 - val_accuracy: 0.7022\r\n",
      "\r\n",
      "Epoch 00066: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 67/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.5049 - accuracy: 0.7425 - val_loss: 0.5429 - val_accuracy: 0.7132\r\n",
      "\r\n",
      "Epoch 00067: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 68/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.4979 - accuracy: 0.7527 - val_loss: 0.5009 - val_accuracy: 0.7610\r\n",
      "\r\n",
      "Epoch 00068: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 69/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4774 - accuracy: 0.7711 - val_loss: 0.5326 - val_accuracy: 0.7500\r\n",
      "\r\n",
      "Epoch 00069: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 70/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.4892 - accuracy: 0.7717 - val_loss: 0.4900 - val_accuracy: 0.7500\r\n",
      "\r\n",
      "Epoch 00070: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 71/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4745 - accuracy: 0.7615 - val_loss: 0.5687 - val_accuracy: 0.7537\r\n",
      "\r\n",
      "Epoch 00071: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 72/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.4722 - accuracy: 0.7853 - val_loss: 0.5411 - val_accuracy: 0.7574\r\n",
      "\r\n",
      "Epoch 00072: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 73/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.4738 - accuracy: 0.7629 - val_loss: 0.5138 - val_accuracy: 0.7426\r\n",
      "\r\n",
      "Epoch 00073: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 74/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4602 - accuracy: 0.7806 - val_loss: 0.5417 - val_accuracy: 0.7132\r\n",
      "\r\n",
      "Epoch 00074: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 75/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4603 - accuracy: 0.7860 - val_loss: 0.5302 - val_accuracy: 0.7353\r\n",
      "\r\n",
      "Epoch 00075: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 76/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.4385 - accuracy: 0.7867 - val_loss: 0.5202 - val_accuracy: 0.7353\r\n",
      "\r\n",
      "Epoch 00076: val_accuracy did not improve from 0.76103\r\n",
      "Epoch 77/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4568 - accuracy: 0.7704 - val_loss: 0.4655 - val_accuracy: 0.7868\r\n",
      "\r\n",
      "Epoch 00077: val_accuracy improved from 0.76103 to 0.78676, saving model to ./vgg16.h5\r\n",
      "Epoch 78/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.4534 - accuracy: 0.7962 - val_loss: 0.5263 - val_accuracy: 0.7279\r\n",
      "\r\n",
      "Epoch 00078: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 79/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.4400 - accuracy: 0.8016 - val_loss: 0.5103 - val_accuracy: 0.7463\r\n",
      "\r\n",
      "Epoch 00079: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 80/150\r\n",
      "184/184 [==============================] - 10s 53ms/step - loss: 0.4458 - accuracy: 0.7745 - val_loss: 0.5426 - val_accuracy: 0.7243\r\n",
      "\r\n",
      "Epoch 00080: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 81/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4280 - accuracy: 0.8030 - val_loss: 0.5541 - val_accuracy: 0.7537\r\n",
      "\r\n",
      "Epoch 00081: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 82/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4150 - accuracy: 0.8077 - val_loss: 0.5012 - val_accuracy: 0.7537\r\n",
      "\r\n",
      "Epoch 00082: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 83/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.4123 - accuracy: 0.8125 - val_loss: 0.4709 - val_accuracy: 0.7794\r\n",
      "\r\n",
      "Epoch 00083: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 84/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.4062 - accuracy: 0.8111 - val_loss: 0.4677 - val_accuracy: 0.7647\r\n",
      "\r\n",
      "Epoch 00084: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 85/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4236 - accuracy: 0.8071 - val_loss: 0.5008 - val_accuracy: 0.7684\r\n",
      "\r\n",
      "Epoch 00085: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 86/150\r\n",
      "184/184 [==============================] - 10s 53ms/step - loss: 0.4128 - accuracy: 0.8111 - val_loss: 0.4793 - val_accuracy: 0.7684\r\n",
      "\r\n",
      "Epoch 00086: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 87/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.4034 - accuracy: 0.8288 - val_loss: 0.5007 - val_accuracy: 0.7684\r\n",
      "\r\n",
      "Epoch 00087: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 88/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.4050 - accuracy: 0.8084 - val_loss: 0.5459 - val_accuracy: 0.7426\r\n",
      "\r\n",
      "Epoch 00088: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 89/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.3951 - accuracy: 0.8179 - val_loss: 0.5449 - val_accuracy: 0.7243\r\n",
      "\r\n",
      "Epoch 00089: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 90/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.3930 - accuracy: 0.8166 - val_loss: 0.6395 - val_accuracy: 0.7022\r\n",
      "\r\n",
      "Epoch 00090: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 91/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.3881 - accuracy: 0.8302 - val_loss: 0.5861 - val_accuracy: 0.7279\r\n",
      "\r\n",
      "Epoch 00091: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 92/150\r\n",
      "184/184 [==============================] - 9s 51ms/step - loss: 0.3939 - accuracy: 0.8166 - val_loss: 0.5278 - val_accuracy: 0.7426\r\n",
      "\r\n",
      "Epoch 00092: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 93/150\r\n",
      "184/184 [==============================] - 10s 54ms/step - loss: 0.3764 - accuracy: 0.8268 - val_loss: 0.5318 - val_accuracy: 0.7390\r\n",
      "\r\n",
      "Epoch 00093: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 94/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.3665 - accuracy: 0.8322 - val_loss: 0.6534 - val_accuracy: 0.7500\r\n",
      "\r\n",
      "Epoch 00094: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 95/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.3803 - accuracy: 0.8295 - val_loss: 0.5630 - val_accuracy: 0.7610\r\n",
      "\r\n",
      "Epoch 00095: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 96/150\r\n",
      "184/184 [==============================] - 10s 52ms/step - loss: 0.3737 - accuracy: 0.8295 - val_loss: 0.6718 - val_accuracy: 0.7316\r\n",
      "\r\n",
      "Epoch 00096: val_accuracy did not improve from 0.78676\r\n",
      "Epoch 97/150\r\n",
      "184/184 [==============================] - 9s 52ms/step - loss: 0.3524 - accuracy: 0.8458 - val_loss: 0.5585 - val_accuracy: 0.7574\r\n",
      "\r\n",
      "Epoch 00097: val_accuracy did not improve from 0.78676\r\n",
      "I0810 12:19:20.227279 139973921728320 train_full.py:192] Kiểm tra trên tập test\r\n",
      "63/63 [==============================] - 3s 46ms/step - loss: 0.5204 - accuracy: 0.7560\r\n"
     ]
    }
   ],
   "source": [
    "!python train_full.py  --batch_size=8\\\n",
    "                        --load_dir='/kaggle/input/model-save/repo/vgg16.h5'\\\n",
    "                        --save_dir='./vgg16.h5'\\\n",
    "                        --num_class=1\\\n",
    "                        --lr_step1=0.0001\\\n",
    "                        --lr_step2=0.00001\\\n",
    "                        --epochs_step1=50\\\n",
    "                        --epochs_step2=100\\\n",
    "                        --image_size=448\\\n",
    "                        --num_channels1=256\\\n",
    "                        --num_channels2=256"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "3ff8821c",
   "metadata": {
    "execution": {
     "iopub.execute_input": "2022-08-10T12:19:25.350887Z",
     "iopub.status.busy": "2022-08-10T12:19:25.350048Z",
     "iopub.status.idle": "2022-08-10T12:31:30.768100Z",
     "shell.execute_reply": "2022-08-10T12:31:30.768908Z"
    },
    "papermill": {
     "duration": 725.542352,
     "end_time": "2022-08-10T12:31:30.769205",
     "exception": false,
     "start_time": "2022-08-10T12:19:25.226853",
     "status": "completed"
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2022-08-10 12:19:26.770358: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/conda/lib\r\n",
      "2022-08-10 12:19:26.770426: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.\r\n",
      "2022-08-10 12:19:30.780475: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set\r\n",
      "2022-08-10 12:19:30.780782: W tensorflow/stream_executor/platform/default/dso_loader.cc:60] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /opt/conda/lib\r\n",
      "2022-08-10 12:19:30.780817: W tensorflow/stream_executor/cuda/cuda_driver.cc:326] failed call to cuInit: UNKNOWN ERROR (303)\r\n",
      "2022-08-10 12:19:30.780847: I tensorflow/stream_executor/cuda/cuda_diagnostics.cc:156] kernel driver does not appear to be running on this host (7b4cc419d731): /proc/driver/nvidia/version does not exist\r\n",
      "2022-08-10 12:19:30.781482: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA\r\n",
      "To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.\r\n",
      "2022-08-10 12:19:30.781846: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set\r\n",
      "2022-08-10 12:19:30.782469: I tensorflow/compiler/jit/xla_gpu_device.cc:99] Not creating XLA devices, tf_xla_enable_xla_devices not set\r\n",
      "2022-08-10 12:19:30.788121: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job worker -> {0 -> 10.0.0.2:8470}\r\n",
      "2022-08-10 12:19:30.788164: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job localhost -> {0 -> localhost:30180}\r\n",
      "2022-08-10 12:19:30.800469: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job worker -> {0 -> 10.0.0.2:8470}\r\n",
      "2022-08-10 12:19:30.800697: I tensorflow/core/distributed_runtime/rpc/grpc_channel.cc:301] Initialize GrpcChannelCache for job localhost -> {0 -> localhost:30180}\r\n",
      "2022-08-10 12:19:30.801385: I tensorflow/core/distributed_runtime/rpc/grpc_server_lib.cc:411] Started server with target: grpc://localhost:30180\r\n",
      "I0810 12:19:30.802511 140229952649024 remote.py:218] Entering into master device scope: /job:worker/replica:0/task:0/device:CPU:0\r\n",
      "I0810 12:19:30.805010 140229952649024 tpu_strategy_util.py:74] Initializing the TPU system: grpc://10.0.0.2:8470\r\n",
      "I0810 12:19:36.711258 140229952649024 tpu_strategy_util.py:109] Clearing out eager caches\r\n",
      "I0810 12:19:36.713155 140229952649024 tpu_strategy_util.py:133] Finished initializing TPU system.\r\n",
      "I0810 12:19:36.715081 140229952649024 tpu_system_metadata.py:159] Found TPU system:\r\n",
      "I0810 12:19:36.715290 140229952649024 tpu_system_metadata.py:160] *** Num TPU Cores: 8\r\n",
      "I0810 12:19:36.715561 140229952649024 tpu_system_metadata.py:161] *** Num TPU Workers: 1\r\n",
      "I0810 12:19:36.715692 140229952649024 tpu_system_metadata.py:163] *** Num TPU Cores Per Worker: 8\r\n",
      "I0810 12:19:36.715766 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:localhost/replica:0/task:0/device:CPU:0, CPU, 0, 0)\r\n",
      "I0810 12:19:36.715958 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:CPU:0, CPU, 0, 0)\r\n",
      "I0810 12:19:36.716031 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:0, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716101 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:1, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716168 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:2, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716235 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:3, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716301 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:4, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716367 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:5, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716430 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:6, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716495 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU:7, TPU, 0, 0)\r\n",
      "I0810 12:19:36.716560 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:TPU_SYSTEM:0, TPU_SYSTEM, 0, 0)\r\n",
      "I0810 12:19:36.716647 140229952649024 tpu_system_metadata.py:165] *** Available Device: _DeviceAttributes(/job:worker/replica:0/task:0/device:XLA_CPU:0, XLA_CPU, 0, 0)\r\n",
      "2022-08-10 12:19:37.084244: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.147823: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.218988: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.286744: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.357257: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.427570: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.495971: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.574839: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.646129: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.762420: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:37.845244: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "I0810 12:19:37.910295 140229952649024 dataset_info.py:358] Load dataset info from gs://kds-0289b529d96a785933aaa9a1abf3af61a5788358dc2bbd021561b2ef/full_images/1.0.0\r\n",
      "2022-08-10 12:19:37.960551: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:38.091879: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:38.166792: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:38.280568: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:38.356511: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:38.423728: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "2022-08-10 12:19:38.495003: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "I0810 12:19:38.567111 140229952649024 dataset_builder.py:351] Reusing dataset full_images (gs://kds-0289b529d96a785933aaa9a1abf3af61a5788358dc2bbd021561b2ef/full_images/1.0.0)\r\n",
      "I0810 12:19:38.567497 140229952649024 logging_logger.py:36] Constructing tf.data.Dataset full_images for split ['train', 'test', 'val'], from gs://kds-0289b529d96a785933aaa9a1abf3af61a5788358dc2bbd021561b2ef/full_images/1.0.0\r\n",
      "2022-08-10 12:19:38.567751: I tensorflow/core/platform/cloud/google_auth_provider.cc:180] Attempting an empty bearer token since no token was retrieved from files, and GCE metadata check was skipped.\r\n",
      "I0810 12:19:39.188281 140229952649024 train_full.py:153] Huấn luyện giai đoạn 1 với các lớp được thêm vào\r\n",
      "Model: \"model_1\"\r\n",
      "_________________________________________________________________\r\n",
      "Layer (type)                 Output Shape              Param #   \r\n",
      "=================================================================\r\n",
      "input_1 (InputLayer)         [(None, 448, 448, 3)]     0         \r\n",
      "_________________________________________________________________\r\n",
      "model (Functional)           (None, None, None, 2048)  23587712  \r\n",
      "_________________________________________________________________\r\n",
      "sequential_3 (Sequential)    (None, 7, 7, 256)         4718848   \r\n",
      "_________________________________________________________________\r\n",
      "sequential_4 (Sequential)    (None, 3, 3, 256)         590080    \r\n",
      "_________________________________________________________________\r\n",
      "global_average_pooling2d (Gl (None, 256)               0         \r\n",
      "_________________________________________________________________\r\n",
      "dense (Dense)                (None, 1)                 257       \r\n",
      "=================================================================\r\n",
      "Total params: 28,896,897\r\n",
      "Trainable params: 28,843,777\r\n",
      "Non-trainable params: 53,120\r\n",
      "_________________________________________________________________\r\n",
      "I0810 12:19:53.796019 140229952649024 train_full.py:165] None\r\n",
      "Epoch 1/50\r\n",
      "184/184 [==============================] - 45s 243ms/step - loss: 0.8529 - accuracy: 0.5340 - val_loss: 0.6571 - val_accuracy: 0.6066\r\n",
      "\r\n",
      "Epoch 00001: val_accuracy improved from -inf to 0.60662, saving model to ./resnet50.h5\r\n",
      "Epoch 2/50\r\n",
      "184/184 [==============================] - 5s 29ms/step - loss: 0.7047 - accuracy: 0.5448 - val_loss: 0.6627 - val_accuracy: 0.5441\r\n",
      "\r\n",
      "Epoch 00002: val_accuracy did not improve from 0.60662\r\n",
      "Epoch 3/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6754 - accuracy: 0.5815 - val_loss: 0.6554 - val_accuracy: 0.5809\r\n",
      "\r\n",
      "Epoch 00003: val_accuracy did not improve from 0.60662\r\n",
      "Epoch 4/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6632 - accuracy: 0.5856 - val_loss: 0.6753 - val_accuracy: 0.5441\r\n",
      "\r\n",
      "Epoch 00004: val_accuracy did not improve from 0.60662\r\n",
      "Epoch 5/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6544 - accuracy: 0.6236 - val_loss: 0.6584 - val_accuracy: 0.5699\r\n",
      "\r\n",
      "Epoch 00005: val_accuracy did not improve from 0.60662\r\n",
      "Epoch 6/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6510 - accuracy: 0.6162 - val_loss: 0.6700 - val_accuracy: 0.5846\r\n",
      "\r\n",
      "Epoch 00006: val_accuracy did not improve from 0.60662\r\n",
      "Epoch 7/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6418 - accuracy: 0.6413 - val_loss: 0.6307 - val_accuracy: 0.6287\r\n",
      "\r\n",
      "Epoch 00007: val_accuracy improved from 0.60662 to 0.62868, saving model to ./resnet50.h5\r\n",
      "Epoch 8/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.6346 - accuracy: 0.6447 - val_loss: 0.6826 - val_accuracy: 0.5441\r\n",
      "\r\n",
      "Epoch 00008: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 9/50\r\n",
      "184/184 [==============================] - 5s 30ms/step - loss: 0.6303 - accuracy: 0.6427 - val_loss: 0.6587 - val_accuracy: 0.6029\r\n",
      "\r\n",
      "Epoch 00009: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 10/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6056 - accuracy: 0.6766 - val_loss: 0.6700 - val_accuracy: 0.5735\r\n",
      "\r\n",
      "Epoch 00010: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 11/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6014 - accuracy: 0.6827 - val_loss: 0.7103 - val_accuracy: 0.5699\r\n",
      "\r\n",
      "Epoch 00011: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 12/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.6069 - accuracy: 0.6692 - val_loss: 0.6435 - val_accuracy: 0.5882\r\n",
      "\r\n",
      "Epoch 00012: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 13/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.5905 - accuracy: 0.6882 - val_loss: 0.6313 - val_accuracy: 0.6250\r\n",
      "\r\n",
      "Epoch 00013: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 14/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5902 - accuracy: 0.6895 - val_loss: 0.7247 - val_accuracy: 0.5846\r\n",
      "\r\n",
      "Epoch 00014: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 15/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5777 - accuracy: 0.6936 - val_loss: 0.6848 - val_accuracy: 0.6066\r\n",
      "\r\n",
      "Epoch 00015: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 16/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.5817 - accuracy: 0.6963 - val_loss: 0.7555 - val_accuracy: 0.5551\r\n",
      "\r\n",
      "Epoch 00016: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 17/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5722 - accuracy: 0.7086 - val_loss: 0.6431 - val_accuracy: 0.5956\r\n",
      "\r\n",
      "Epoch 00017: val_accuracy did not improve from 0.62868\r\n",
      "Epoch 18/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.5696 - accuracy: 0.7024 - val_loss: 0.6170 - val_accuracy: 0.6434\r\n",
      "\r\n",
      "Epoch 00018: val_accuracy improved from 0.62868 to 0.64338, saving model to ./resnet50.h5\r\n",
      "Epoch 19/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5514 - accuracy: 0.7188 - val_loss: 0.9368 - val_accuracy: 0.5221\r\n",
      "\r\n",
      "Epoch 00019: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 20/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.5430 - accuracy: 0.7351 - val_loss: 0.9660 - val_accuracy: 0.5257\r\n",
      "\r\n",
      "Epoch 00020: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 21/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5548 - accuracy: 0.7249 - val_loss: 0.7075 - val_accuracy: 0.5956\r\n",
      "\r\n",
      "Epoch 00021: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 22/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5622 - accuracy: 0.7004 - val_loss: 0.8024 - val_accuracy: 0.5404\r\n",
      "\r\n",
      "Epoch 00022: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 23/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5373 - accuracy: 0.7351 - val_loss: 0.8221 - val_accuracy: 0.5772\r\n",
      "\r\n",
      "Epoch 00023: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 24/50\r\n",
      "184/184 [==============================] - 5s 30ms/step - loss: 0.5244 - accuracy: 0.7289 - val_loss: 0.7574 - val_accuracy: 0.6176\r\n",
      "\r\n",
      "Epoch 00024: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 25/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5312 - accuracy: 0.7344 - val_loss: 0.6957 - val_accuracy: 0.6397\r\n",
      "\r\n",
      "Epoch 00025: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 26/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5110 - accuracy: 0.7432 - val_loss: 0.7713 - val_accuracy: 0.5993\r\n",
      "\r\n",
      "Epoch 00026: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 27/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5142 - accuracy: 0.7371 - val_loss: 0.7765 - val_accuracy: 0.6103\r\n",
      "\r\n",
      "Epoch 00027: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 28/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.4921 - accuracy: 0.7541 - val_loss: 0.9812 - val_accuracy: 0.5294\r\n",
      "\r\n",
      "Epoch 00028: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 29/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.5058 - accuracy: 0.7500 - val_loss: 0.7898 - val_accuracy: 0.5919\r\n",
      "\r\n",
      "Epoch 00029: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 30/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5024 - accuracy: 0.7507 - val_loss: 0.8522 - val_accuracy: 0.5772\r\n",
      "\r\n",
      "Epoch 00030: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 31/50\r\n",
      "184/184 [==============================] - 5s 29ms/step - loss: 0.5040 - accuracy: 0.7500 - val_loss: 0.7276 - val_accuracy: 0.5956\r\n",
      "\r\n",
      "Epoch 00031: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 32/50\r\n",
      "184/184 [==============================] - 5s 28ms/step - loss: 0.4931 - accuracy: 0.7520 - val_loss: 0.7619 - val_accuracy: 0.5956\r\n",
      "\r\n",
      "Epoch 00032: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 33/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.4785 - accuracy: 0.7568 - val_loss: 0.8038 - val_accuracy: 0.6140\r\n",
      "\r\n",
      "Epoch 00033: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 34/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.5003 - accuracy: 0.7582 - val_loss: 0.9205 - val_accuracy: 0.5772\r\n",
      "\r\n",
      "Epoch 00034: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 35/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.4774 - accuracy: 0.7704 - val_loss: 1.0003 - val_accuracy: 0.5846\r\n",
      "\r\n",
      "Epoch 00035: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 36/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.4826 - accuracy: 0.7520 - val_loss: 0.9268 - val_accuracy: 0.5772\r\n",
      "\r\n",
      "Epoch 00036: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 37/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.4618 - accuracy: 0.7683 - val_loss: 0.9319 - val_accuracy: 0.6029\r\n",
      "\r\n",
      "Epoch 00037: val_accuracy did not improve from 0.64338\r\n",
      "Epoch 38/50\r\n",
      "184/184 [==============================] - 5s 27ms/step - loss: 0.4750 - accuracy: 0.7656 - val_loss: 1.1106 - val_accuracy: 0.5441\r\n",
      "\r\n",
      "Epoch 00038: val_accuracy did not improve from 0.64338\r\n",
      "I0810 12:24:10.988484 140229952649024 train_full.py:176] Mở đóng băng mô hình, huấn luyện giai đoạn 2\r\n",
      "Epoch 38/150\r\n",
      "184/184 [==============================] - 89s 483ms/step - loss: 0.6581 - accuracy: 0.6311 - val_loss: 0.6891 - val_accuracy: 0.6581\r\n",
      "\r\n",
      "Epoch 00038: val_accuracy improved from 0.64338 to 0.65809, saving model to ./resnet50.h5\r\n",
      "Epoch 39/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.6183 - accuracy: 0.6576 - val_loss: 0.7116 - val_accuracy: 0.6434\r\n",
      "\r\n",
      "Epoch 00039: val_accuracy did not improve from 0.65809\r\n",
      "Epoch 40/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.5904 - accuracy: 0.6698 - val_loss: 0.6708 - val_accuracy: 0.6213\r\n",
      "\r\n",
      "Epoch 00040: val_accuracy did not improve from 0.65809\r\n",
      "Epoch 41/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.5690 - accuracy: 0.7038 - val_loss: 0.7459 - val_accuracy: 0.6544\r\n",
      "\r\n",
      "Epoch 00041: val_accuracy did not improve from 0.65809\r\n",
      "Epoch 42/150\r\n",
      "184/184 [==============================] - 8s 42ms/step - loss: 0.5568 - accuracy: 0.7004 - val_loss: 0.8150 - val_accuracy: 0.6544\r\n",
      "\r\n",
      "Epoch 00042: val_accuracy did not improve from 0.65809\r\n",
      "Epoch 43/150\r\n",
      "184/184 [==============================] - 8s 42ms/step - loss: 0.5515 - accuracy: 0.7283 - val_loss: 0.6280 - val_accuracy: 0.6618\r\n",
      "\r\n",
      "Epoch 00043: val_accuracy improved from 0.65809 to 0.66176, saving model to ./resnet50.h5\r\n",
      "Epoch 44/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.5248 - accuracy: 0.7364 - val_loss: 0.6740 - val_accuracy: 0.6654\r\n",
      "\r\n",
      "Epoch 00044: val_accuracy improved from 0.66176 to 0.66544, saving model to ./resnet50.h5\r\n",
      "Epoch 45/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4980 - accuracy: 0.7554 - val_loss: 0.7587 - val_accuracy: 0.6434\r\n",
      "\r\n",
      "Epoch 00045: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 46/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4871 - accuracy: 0.7452 - val_loss: 0.7245 - val_accuracy: 0.6287\r\n",
      "\r\n",
      "Epoch 00046: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 47/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4970 - accuracy: 0.7493 - val_loss: 0.6872 - val_accuracy: 0.6507\r\n",
      "\r\n",
      "Epoch 00047: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 48/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4750 - accuracy: 0.7697 - val_loss: 0.7306 - val_accuracy: 0.6397\r\n",
      "\r\n",
      "Epoch 00048: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 49/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4474 - accuracy: 0.7846 - val_loss: 0.7001 - val_accuracy: 0.6507\r\n",
      "\r\n",
      "Epoch 00049: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 50/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4641 - accuracy: 0.7840 - val_loss: 0.6589 - val_accuracy: 0.6654\r\n",
      "\r\n",
      "Epoch 00050: val_accuracy did not improve from 0.66544\r\n",
      "Epoch 51/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4435 - accuracy: 0.7928 - val_loss: 0.6144 - val_accuracy: 0.6875\r\n",
      "\r\n",
      "Epoch 00051: val_accuracy improved from 0.66544 to 0.68750, saving model to ./resnet50.h5\r\n",
      "Epoch 52/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4416 - accuracy: 0.7935 - val_loss: 0.6268 - val_accuracy: 0.7096\r\n",
      "\r\n",
      "Epoch 00052: val_accuracy improved from 0.68750 to 0.70956, saving model to ./resnet50.h5\r\n",
      "Epoch 53/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4080 - accuracy: 0.8016 - val_loss: 0.6028 - val_accuracy: 0.6949\r\n",
      "\r\n",
      "Epoch 00053: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 54/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.4240 - accuracy: 0.7955 - val_loss: 0.6207 - val_accuracy: 0.7022\r\n",
      "\r\n",
      "Epoch 00054: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 55/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3979 - accuracy: 0.8254 - val_loss: 0.6476 - val_accuracy: 0.6838\r\n",
      "\r\n",
      "Epoch 00055: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 56/150\r\n",
      "184/184 [==============================] - 9s 47ms/step - loss: 0.3923 - accuracy: 0.8159 - val_loss: 0.6727 - val_accuracy: 0.6728\r\n",
      "\r\n",
      "Epoch 00056: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 57/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3938 - accuracy: 0.8145 - val_loss: 0.6993 - val_accuracy: 0.6875\r\n",
      "\r\n",
      "Epoch 00057: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 58/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3907 - accuracy: 0.8302 - val_loss: 0.7025 - val_accuracy: 0.6912\r\n",
      "\r\n",
      "Epoch 00058: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 59/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3617 - accuracy: 0.8485 - val_loss: 0.6999 - val_accuracy: 0.6581\r\n",
      "\r\n",
      "Epoch 00059: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 60/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3594 - accuracy: 0.8383 - val_loss: 0.7677 - val_accuracy: 0.6728\r\n",
      "\r\n",
      "Epoch 00060: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 61/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3553 - accuracy: 0.8390 - val_loss: 0.6648 - val_accuracy: 0.7022\r\n",
      "\r\n",
      "Epoch 00061: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 62/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3629 - accuracy: 0.8383 - val_loss: 0.6867 - val_accuracy: 0.6838\r\n",
      "\r\n",
      "Epoch 00062: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 63/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3473 - accuracy: 0.8390 - val_loss: 0.7958 - val_accuracy: 0.6765\r\n",
      "\r\n",
      "Epoch 00063: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 64/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3300 - accuracy: 0.8689 - val_loss: 0.7558 - val_accuracy: 0.6765\r\n",
      "\r\n",
      "Epoch 00064: val_accuracy did not improve from 0.70956\r\n",
      "Epoch 65/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.2901 - accuracy: 0.8859 - val_loss: 0.7338 - val_accuracy: 0.7243\r\n",
      "\r\n",
      "Epoch 00065: val_accuracy improved from 0.70956 to 0.72426, saving model to ./resnet50.h5\r\n",
      "Epoch 66/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3234 - accuracy: 0.8465 - val_loss: 0.7357 - val_accuracy: 0.7096\r\n",
      "\r\n",
      "Epoch 00066: val_accuracy did not improve from 0.72426\r\n",
      "Epoch 67/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3026 - accuracy: 0.8716 - val_loss: 0.7580 - val_accuracy: 0.7059\r\n",
      "\r\n",
      "Epoch 00067: val_accuracy did not improve from 0.72426\r\n",
      "Epoch 68/150\r\n",
      "184/184 [==============================] - 9s 47ms/step - loss: 0.3138 - accuracy: 0.8594 - val_loss: 0.7356 - val_accuracy: 0.7022\r\n",
      "\r\n",
      "Epoch 00068: val_accuracy did not improve from 0.72426\r\n",
      "Epoch 69/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.2775 - accuracy: 0.8832 - val_loss: 0.7782 - val_accuracy: 0.7243\r\n",
      "\r\n",
      "Epoch 00069: val_accuracy did not improve from 0.72426\r\n",
      "Epoch 70/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.3029 - accuracy: 0.8655 - val_loss: 0.7312 - val_accuracy: 0.7059\r\n",
      "\r\n",
      "Epoch 00070: val_accuracy did not improve from 0.72426\r\n",
      "Epoch 71/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.2914 - accuracy: 0.8743 - val_loss: 0.7467 - val_accuracy: 0.7243\r\n",
      "\r\n",
      "Epoch 00071: val_accuracy did not improve from 0.72426\r\n",
      "Epoch 72/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.2699 - accuracy: 0.8872 - val_loss: 0.7848 - val_accuracy: 0.6985\r\n",
      "\r\n",
      "Epoch 00072: val_accuracy did not improve from 0.72426\r\n",
      "Epoch 73/150\r\n",
      "184/184 [==============================] - 8s 43ms/step - loss: 0.2683 - accuracy: 0.8879 - val_loss: 0.7854 - val_accuracy: 0.7353\r\n",
      "\r\n",
      "Epoch 00073: val_accuracy improved from 0.72426 to 0.73529, saving model to ./resnet50.h5\r\n",
      "I0810 12:31:24.165809 140229952649024 train_full.py:192] Kiểm tra trên tập test\r\n",
      "63/63 [==============================] - 2s 33ms/step - loss: 0.8214 - accuracy: 0.6964\r\n"
     ]
    }
   ],
   "source": [
    "!python train_full.py  --batch_size=8\\\n",
    "                        --load_dir='/kaggle/input/model-save/repo/resnet50.h5'\\\n",
    "                        --save_dir='./resnet50.h5'\\\n",
    "                        --num_class=1\\\n",
    "                        --lr_step1=0.0001\\\n",
    "                        --lr_step2=0.00001\\\n",
    "                        --epochs_step1=50\\\n",
    "                        --epochs_step2=100\\\n",
    "                        --image_size=448\\\n",
    "                        --num_channels1=256\\\n",
    "                        --num_channels2=256"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  },
  "papermill": {
   "default_parameters": {},
   "duration": 1601.001964,
   "end_time": "2022-08-10T12:31:31.653769",
   "environment_variables": {},
   "exception": null,
   "input_path": "__notebook__.ipynb",
   "output_path": "__notebook__.ipynb",
   "parameters": {},
   "start_time": "2022-08-10T12:04:50.651805",
   "version": "2.3.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
